### ISSUE TITLE
Assertion failure in GuiEffectVisitor.visitMethodInvocation()

### ISSUE THREAD

**Issue by dbrosoft** [typetools/checker-framework#689]
Save [Issue689.zip](https://github.com/typetools/checker-framework/files/239799/Issue689.zip) and proceed as follows:
```sh
$ unzip Issue689.zip
$ cd Issue689
$ java -jar $CHECKERFRAMEWORK/checker/dist/checker.jar -processor guieffect -AprintErrorStack java/util/AbstractList.java
```
It appears (based on experiments commenting out method code) that the problematic method in `AbstractList.java` is `listIterator()` in class `SubList`.  The stack trace of the assertion failure is provided in `error.txt`.

**Comment by csgordon** [comment]
It looks like ```TreeUtils.enclosingMethod(getCurrentPath())``` may be misbehaving when called in the middle of a visit to a field initializer.

The specific code the checker is tripping on is in the anonymous inner class inside SubList that implements its iterator, starting at line 715 of the file included in the report:
```java
        return new ListIterator<E>() {
            private final ListIterator<E> i = l.listIterator(index+offset);
...
```
The compiler of synthesizes a constructor for the anonymous class, which is visited by the ```GuiEffectVisitor```.  On entry to that method, the ```MethodTree``` argument to ```visitMethod``` looks as expected, we push it on a stack of method trees, and visit the body.

When we get to the call in that field initializer, we look up the enclosing MethodTree via ```TreeUtils.enclosingMethod(getCurrentPath())```.  I'm not sure what happens, but when I print the resulting MethodTree, I see instead the tree for the enclosing **anonymous class**, not the method.  I would expect the call above to produce the enclosing synthesized method... @wmdietl?

But that's not the whole story.  I mentioned that I push the method tree onto a stack when we enter it; when we get to the initializer, we've pushed an extra null on the stack...

Back at this in a bit...

**Comment by mernst** [comment]
Closed by PR #729
